WIP: Flash Attention implementation (forward + backward)

FSSRepo opened the pull request:
Previus work: [llama.cpp#778](https://github.com/ggerganov/llama.cpp/pull/778)

Previously, the initiative to implement Flash Attention to improve inference performance in llama.cpp had already been introduced. However, it was assumed that this approach would yield the expected results on the CPU, for that reason, it was discarded, and no further follow-up was given.

Flash Attention is actually designed to enhance GPU resource utilization through the use of tensor cores and shared memory, which is 30 times faster than global memory (VRAM), reducing unnecessary readings and writings.

Implementing this algorithm is particularly challenging because it requires taking into account hardware limitations and conflicts that can degrade performance, even when everything is expected to be fine (a.k.a. shared memory banks conflicts).

Tasks to be carried out during the execution of this project:

- [x] Analyzing and understanding concisely how Flash Attention works. Even if it's just creating a C++ code that replicates the exact functionality without much complexity. From this base, we move on to parallelize and optimize.
- [ ] Optimize and support different models.
- [ ] Create the backward kernel for training.

### Run test:

```bash
# get the code
git clone https://github.com/Pints-App/llama.cpp.git
cd llama.cpp

# build
mkdir build
cd build
cmake .. -DLLAMA_CUBLAS=ON
cmake --build . --config Release

# Run test (get result)
bin/test-flash-attention

# Run test (compare backends)
bin/test-flash-attention comp
```

FSSRepo committed:
cuda: add flash attention + test

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+#define CUDA_FLASH_ATTENTION_BLOCK_SIZE 256
+
+template<int block_size>
+static __global__ void flash_attn_f32(const float* q, const float* k,const float* v, float* dst, float kq_scale,
```
Add `__restrict__` to pointers, see https://github.com/ggerganov/llama.cpp/pull/2140 . On Ampere I saw comparatively little difference but adding it shouldn't hurt.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    const int head = blockIdx.x / seq_len;
+    const int head_size = d_head * seq_len;
+    const int s = blockIdx.x % seq_len;
+    const int tid = threadIdx.x;
```
Explicitly setting an int with `threadIdx.x` moves it to a register. Once you get to optimizing performance, check whether or not register pressure is a limitation. If it is I would advise you to replace `tid` with `threadIdx.x`.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    const int s = blockIdx.x % seq_len;
+    const int tid = threadIdx.x;
+
+    extern __shared__  char work_data[];
```
Do not use a generic name like `work_data` or there will be conflicts with other kernels.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    float* warp_data = (float*)(work_data + seq_len * sizeof(float));
+
+    // QK^T
+    for(int is = tid; is < seq_len; is += block_size) {
```
Potential performance optimization: add the option to notify the compiler of `seq_len` via template and use `#pragma unroll`. See https://github.com/ggerganov/llama.cpp/pull/4742/files .

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    const int tid = threadIdx.x;
+
+    extern __shared__  char work_data[];
+    float* S = (float*)work_data; // theorical sequent length: 12848, due memory per block limit
+    float* warp_data = (float*)(work_data + seq_len * sizeof(float));
```
Tensor cores can handle FP16, BF16, and TF32 but **not** regular floats. So computing data as FP32 is fine for a test implementation to ensure that results are correct but for the actual production implementation you will need to use something else.

Also I don't understand what that comment is supposed to tell me.

FSSRepo replied:
At the moment, it is set up like this. I know that currently the best approach would be to use tensor cores for performing GEMM in this kernel. However, I wonder if I should then convert from f32 to f16 before executing the flash attention kernel and utilize tensor cores to accelerate GEMM.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    // QK^T
+    for(int is = tid; is < seq_len; is += block_size) {
+        S[is] = 0.0f;
+        int key_offset = is * d_head + head * head_size;
+        int query_offset = s * d_head + head * head_size;
```
```suggestion
        const int key_offset = is * d_head + head * head_size;
        const int query_offset = s * d_head + head * head_size;
```

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+        S[is] = 0.0f;
+        int key_offset = is * d_head + head * head_size;
+        int query_offset = s * d_head + head * head_size;
+        for(int d = 0; d < d_head; d++) {
```
Make `d_head` a template parameter and use `#pragma unroll`. I very much expect this to be faster.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+        int key_offset = is * d_head + head * head_size;
+        int query_offset = s * d_head + head * head_size;
+        for(int d = 0; d < d_head; d++) {
+                S[is] += k[key_offset + d] * q[query_offset + d];
```
Copying data from VRAM to SRAM is more efficient using [asynchronous data copies](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#asynchronous-data-copies). On Ampere or newer it is both faster and reduces register pressure. It also allows you to load data and do computations in parallel.  This optimization can be added later though. I will soon push a prototype to https://github.com/ggerganov/llama.cpp/pull/4801 which (I think) is close to optimal.

FSSRepo replied:
From CUDA Documentation:

> memcpy_async is a group-wide collective memcpy that utilizes hardware accelerated support for nonblocking memory transactions from global to shared memory. Given a set of threads named in the
group, memcpy_async will move specified amount of bytes or elements of the input type through a
single pipeline stage. Additionally for achieving best performance when using the memcpy_async API,
an alignment of 16 bytes for both shared memory and global memory is required. It is important to
note that while this is a memcpy in the general case, it is only asynchronous if the source is global
memory and the destination is shared memory and both can be addressed with 16, 8, or 4 byte alignments. Asynchronously copied data should only be read following a call to wait or wait_prior which
signals that the corresponding stage has completed moving data to shared memory.

I will try to use `cuda::memcpy_async` to transfer the data to shared memory for both the query and key matrices in order to perform matrix multiplications.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+
+    float max_val = -INFINITY;
+    // get the max
+    for(int is = tid; is < seq_len; is += block_size) {
```
```suggestion
    for(int is0 = 0; is0 < seq_len; is0 += block_size) {
    const int is = is0 + tid;
    
    if (is >= seq_len) {
        break;
    }
```
This should be faster.
Also `#pragma unroll`, see above.

ggerganov committed:
ggml : add ggml_flash_attn_ext API

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+
+    // softmax(QK^T)
+    float sum = 0.0f;
+    for(int is = tid; is < seq_len;is += block_size) {
```
Same as above.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    }
+
+    float inv_sum = 1.0f / sum;
+    for(int is = tid; is < seq_len; is += block_size) {
```
Same as above.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+
+    __syncthreads();
+    // softmax(QK^T)V
+    for (int d = tid; d < d_head; d += block_size) {
```
Same as above.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
+    for (int d = tid; d < d_head; d += block_size) {
+        int dst_index = d + s * d_head + head * head_size;
+        int value_offset = d * seq_len +   head * head_size;
+        dst[dst_index] = 0.0f;
+        for(int ic = 0; ic < seq_len; ic++) {
+                dst[dst_index] += v[value_offset + ic] * S[ic];
+        }
```
```suggestion
        float tmp= 0.0f;
        for(int ic = 0; ic < seq_len; ic++) {
                tmp += v[value_offset + ic] * S[ic];
        }
        dst[dst_index] = tmp;
```
The way you're currently doing it repeatedly writes to global memory which is slow. It is faster to instead calculate the result in a register which is on-chip and even faster than SRAM and to only write to global memory at the end.

Also This loop would probably benefit from unrolling.

JohannesGaessler commented:
>Create the backward kernel for training.

Currently I don't think llama.cpp has good support for training at all. I want to work on it at some point but implementing a backwards pass currently doesn't make sense I think.

JohannesGaessler commented:
>Optimize and support different models.

In terms of optimization it will rather need to be done for different GPU + batch size combinations. Fundamentally you want to use large SRAM tiles because that increases arithmetic intensity: you can do more computations per load from global memory. However, at the same time you want to split a computing task into as many blocks as possible since this reduces so-called tail effects where the last wave of a kernel does not have enough blocks to keep all of the streaming multiprocessors on a GPU busy. So for optimizing performance you will need to find a balance between these two problems which is affected by batch size because the number of blocks that the GPU can work on should be proportional to batch size.

What I think would be the best solution is to compile the kernel multiple times with different template parameters and to then choose the best implementation at runtime based on the number of streaming multiprocessors on the GPU and teh shapes of the matrices. I am currently working on essentially this exact problem in https://github.com/ggerganov/llama.cpp/pull/4801 (but for regular matrix multiplication).

JohannesGaessler commented:
Since the target architectures are Ampere and Ada Lovelace consider that copying data from VRAM to SRAM is more efficient using [asynchronous data copies](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#asynchronous-data-copies). This is both faster and reduces register pressure. It also allows you to load data and do computations in parallel.  This optimization can be added later though. I will soon push a prototype to https://github.com/ggerganov/llama.cpp/pull/4801 which (I think) is close to optimal.

JohannesGaessler commented:
The installation instructions seem to be missing `cmake .. -DLLAMA_CUBLAS=ON`. In any case, I cannot compile the project. I get the following error message:

<details>

```

/home/johannesg/Projects/llama.cpp/build [git::flash-attn-cuda *] [johannesg@johannes-pc] [17:19]
> cmake --build . --config Release
[  5%] Built target ggml
[  6%] Built target ggml_static
[  8%] Built target llama
[ 10%] Built target build_info
[ 16%] Built target common
[ 17%] Built target test-quantize-fns
[ 19%] Built target test-quantize-perf
[ 21%] Built target test-sampling
[ 23%] Built target test-tokenizer-0-llama
[ 25%] Built target test-tokenizer-0-falcon
[ 27%] Built target test-tokenizer-1-llama
[ 29%] Built target test-tokenizer-1-bpe
[ 31%] Built target test-grammar-parser
[ 32%] Building CXX object tests/CMakeFiles/test-llama-grammar.dir/test-llama-grammar.cpp.o
[ 33%] Linking CXX executable ../bin/test-llama-grammar
[ 33%] Built target test-llama-grammar
[ 34%] Building CXX object tests/CMakeFiles/test-grad0.dir/test-grad0.cpp.o
[ 35%] Linking CXX executable ../bin/test-grad0
[ 35%] Built target test-grad0
[ 36%] Building CXX object tests/CMakeFiles/test-backend-ops.dir/test-backend-ops.cpp.o
[ 37%] Linking CXX executable ../bin/test-backend-ops
[ 37%] Built target test-backend-ops
[ 38%] Building CXX object tests/CMakeFiles/test-rope.dir/test-rope.cpp.o
[ 39%] Linking CXX executable ../bin/test-rope
[ 39%] Built target test-rope
[ 40%] Building CXX object tests/CMakeFiles/test-flash-attention.dir/test-flash-attention.cpp.o
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:5: warning: "GGML_USE_CUBLAS" redefined
    5 | #define GGML_USE_CUBLAS
      | 
<command-line>: note: this is the location of the previous definition
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: In function ‘bool isinf_or_max(float)’:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:74:34: error: ‘FLT_MAX’ was not declared in this scope
   74 |     return std::isinf(f) || f == FLT_MAX || f == -FLT_MAX;
      |                                  ^~~~~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:19:1: note: ‘FLT_MAX’ is defined in header ‘<cfloat>’; did you forget to ‘#include <cfloat>’?
   18 | #include <vector>
  +++ |+#include <cfloat>
   19 | 
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: At global scope:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:93:6: warning: no previous declaration for ‘void ggml_tensor_set_f32(ggml_tensor*, float, int, int, int, int)’ [-Wmissing-declarations]
   93 | void ggml_tensor_set_f32(struct ggml_tensor* tensor, float value, int l, int k = 0, int j = 0, int i = 0) {
      |      ^~~~~~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:98:7: warning: no previous declaration for ‘float ggml_tensor_get_f32(const ggml_tensor*, int, int, int, int)’ [-Wmissing-declarations]
   98 | float ggml_tensor_get_f32(const ggml_tensor* tensor, int l, int k = 0, int j = 0, int i = 0) {
      |       ^~~~~~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:103:6: warning: no previous declaration for ‘void load_model(test_model&, bool)’ [-Wmissing-declarations]
  103 | void load_model(test_model & model, bool use_gpu = false) {
      |      ^~~~~~~~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: In function ‘void load_model(test_model&, bool)’:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:146:44: warning: ‘double ggml_type_sizef(ggml_type)’ is deprecated: use ggml_row_size() instead [-Wdeprecated-declarations]
  146 |         buffer_size += 30 * ggml_type_sizef(GGML_TYPE_F32); // tensor q
      |                             ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
In file included from /home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:1:
/home/johannesg/Projects/llama.cpp/./ggml.h:672:21: note: declared here
  672 |     GGML_API double ggml_type_sizef(enum ggml_type type), // ggml_type_size()/ggml_blck_size() as float
      |                     ^~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/./ggml.h:202:41: note: in definition of macro ‘GGML_DEPRECATED’
  202 | #    define GGML_DEPRECATED(func, hint) func __attribute__((deprecated(hint)))
      |                                         ^~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:147:44: warning: ‘double ggml_type_sizef(ggml_type)’ is deprecated: use ggml_row_size() instead [-Wdeprecated-declarations]
  147 |         buffer_size += 24 * ggml_type_sizef(GGML_TYPE_F32); // tensor k
      |                             ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/./ggml.h:672:21: note: declared here
  672 |     GGML_API double ggml_type_sizef(enum ggml_type type), // ggml_type_size()/ggml_blck_size() as float
      |                     ^~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/./ggml.h:202:41: note: in definition of macro ‘GGML_DEPRECATED’
  202 | #    define GGML_DEPRECATED(func, hint) func __attribute__((deprecated(hint)))
      |                                         ^~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:148:44: warning: ‘double ggml_type_sizef(ggml_type)’ is deprecated: use ggml_row_size() instead [-Wdeprecated-declarations]
  148 |         buffer_size += 24 * ggml_type_sizef(GGML_TYPE_F32); // tensor v
      |                             ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/./ggml.h:672:21: note: declared here
  672 |     GGML_API double ggml_type_sizef(enum ggml_type type), // ggml_type_size()/ggml_blck_size() as float
      |                     ^~~~~~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/./ggml.h:202:41: note: in definition of macro ‘GGML_DEPRECATED’
  202 | #    define GGML_DEPRECATED(func, hint) func __attribute__((deprecated(hint)))
      |                                         ^~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: At global scope:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:203:22: warning: no previous declaration for ‘ggml_cgraph* build_graph(const test_model&, ggml_allocr*)’ [-Wmissing-declarations]
  203 | struct ggml_cgraph * build_graph(const test_model& model, struct ggml_allocr * allocr) {
      |                      ^~~~~~~~~~~
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: In function ‘ggml_cgraph* build_graph(const test_model&, ggml_allocr*)’:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:203:80: warning: unused parameter ‘allocr’ [-Wunused-parameter]
  203 | build_graph(const test_model& model, struct ggml_allocr * allocr) {
      |                                      ~~~~~~~~~~~~~~~~~~~~~^~~~~~

/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp: At global scope:
/home/johannesg/Projects/llama.cpp/tests/test-flash-attention.cpp:226:21: warning: no previous declaration for ‘ggml_tensor* compute_graph(const test_model&, ggml_backend_t, ggml_allocr*, bool)’ [-Wmissing-declarations]
  226 | struct ggml_tensor* compute_graph(const test_model & model, ggml_backend_t backend_cpu, struct ggml_allocr * allocr, bool compare_backends) {
      |                     ^~~~~~~~~~~~~
make[2]: *** [tests/CMakeFiles/test-flash-attention.dir/build.make:76: tests/CMakeFiles/test-flash-attention.dir/test-flash-attention.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1929: tests/CMakeFiles/test-flash-attention.dir/all] Error 2


```

</details>

JohannesGaessler commented:
As hinted at by the error message, it can be fixed with `#include <cfloat>`.

FSSRepo committed:
fix compilation

ggerganov committed:
ggml : fix GQA support in ggml_flash_attn_ext

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

FSSRepo committed:
apply suggestions

ggerganov committed:
Merge branch 'master' into gg/flash-attn

ggerganov committed:
ggml : online attention (CPU)

ggerganov committed:
metal : initial implementation

ggerganov committed:
metal : f16 precision

ggerganov committed:
metal : reduce branches

ggerganov committed:
metal : specialize for head size

ggerganov committed:
wip : 8 rows per simd group

ggerganov committed:
wip : 4 rows per simd group

ggerganov committed:
wip : template for rows per warp

ggerganov committed:
metal : parallelize across KV size

ggerganov committed:
metal : parallel reduce across heads

ggerganov committed:
metal : efficient flash_attn_f16 implementation

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

FSSRepo committed:
cuda: port metal version flash_attn_ext

FSSRepo committed:
fix equivalent fp16 math functions, compiler error 'undefined'

FSSRepo commented:
@JohannesGaessler I'm trying to do a port of the implementation that Georgi did, but I'm running into issues where CUDA doesn't support fp16 operators. I need to use the functions from cuda_fp16.h, but when I compile, I get that they are undefined.

JohannesGaessler commented:
The issue is that cmake by default compiles device code for compute capabilities 5.2, 6.1, and 7.0. You can fix it for now by compiling with `-DLLAMA_CUDA_F16=ON`.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+    const int head_size = head_dim * seq_len;
+    const int s = blockIdx.x % seq_len;
+
+    extern __shared__  char shmem__[];
```
As (I think) I said before, generic names like `shmem__` will conflict between kernels. Use a name like `data_flash_attn_f32` instead.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+        const int key_offset = is * head_dim + head * head_size;
+        const int query_offset = s * head_dim + head * head_size;
+
+        S[is] = 0.0f;
+        for(int d = 0; d < head_dim; d++) {
+                S[is] += k[key_offset + d] * q[query_offset + d];
+        }
+        S[is] *= kq_scale;
```
What you're doing here is repeatedly reading from and writing to shared memory. It should be faster to calculate the result in a register (just a regular declaration like `float S_tmp = 0.0f;`) and to then only write it to shared memory once at the end.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+            break;
+        }
+
+        S[is] = expf(S[is] - max_val);
+        sum += S[is];
```
More unneeded shared memory reads/writes.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+    __syncthreads();
+
+    // softmax(QK^T)V
+    for (int d = threadIdx.x; d < head_dim; d += block_size) {
```
This loop should be faster if rewritten in the same way as the ones above with `d = d0 + threadIdx.x;`.

JohannesGaessler commented:
Current performance level on my system: 

| GPU      | Model         | Test   |   t/s master |   t/s flash-attn-cuda |   Speedup |
|:---------|:--------------|:-------|-------------:|----------------------:|----------:|
| RTX 3090 | llama 7B Q8_0 | pp512  |      3871.86 |               2402.47 |      0.62 |
| RTX 3090 | llama 7B Q8_0 | tg128  |        85.61 |                 85.65 |      1.00 |

JohannesGaessler commented:
Based on data from NSight Systems this kernel is currently too slow by a factor of ~10.6 to break even with master. 

I would really advise you to get familiar with NSight Compute. You can immediately see that with the current implementation only 16 blocks are being launched for LLaMA 2 7b. Meanwhile my RTX 3090 has 82 streaming multiprocessors which means that 62 of those SMs are idling and the GPU utilization is capped to ~25%. On an RTX 4090 or A100 this is even worse because they have even more SMs. 

Also 1024 is much too large for a matrix multiplication kernel. There are only 65536 registers available per SM and with a block size that large it limits you to 64 registers per thread which is really not a lot. But then again it seems the compiler is assigning only 40 registers to each thread so it may make sense to specify `__launch_bounds__(<BLOCK_SIZE>, 1)` to let the compiler know that you intend to run only a single block of this kernel per SM. Again, this is something that you can see in NSight Compute.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+    half y;
+    half z;
+    half w;
+};
+
+__device__ half4 make_half4(half x) {
+    half4 t;
+    t.x = x; t.y = x; t.z = x; t.w = x;
+    return t;
+}
+
+__device__ half4 __h4fma(half4 a, half b, half4 c) {
+    half4 t;
+    t.x = __hfma(a.x, b, c.x); t.y = __hfma(a.y, b, c.y); t.z = __hfma(a.z, b, c.z); t.w = __hfma(a.w, b, c.w);
+    return t;
+}
+
+__device__ half4 __h4fma(half4 a, half4 b, half4 c) {
+    half4 t;
+    t.x = __hfma(a.x, b.x, c.x); t.y = __hfma(a.y, b.y, c.y); t.z = __hfma(a.z, b.z, c.z); t.w = __hfma(a.w, b.w, c.w);
+    return t;
+}
+
+__device__ half4 __h4mul(half4 a, half b) {
+    half4 t;
+    t.x = __hmul(a.x, b); t.y = __hmul(a.y, b); t.z =__hmul(a.z, b); t.w =__hmul(a.w, b);
+    return t;
+}
+
+__device__ half4 __h4mul(half4 a, half4 b) {
+    half4 t;
+    t.x = __hmul(a.x, b.x); t.y = __hmul(a.y, b.y); t.z =__hmul(a.z, b.z); t.w =__hmul(a.w, b.w);
+    return t;
+}
+
+__device__ half4 __h4div(half4 a, half b) {
+    half4 t;
+    t.x = __hdiv(a.x, b); t.y = __hdiv(a.y, b); t.z =__hdiv(a.z, b); t.w =__hdiv(a.w, b);
+    return t;
+}
```
Rewrite these to use half2 values and arithmetic instead of half values and arithmetic. You can store 2 half values per register but the current code essentially takes one of those registers, extracts one of the half values, does some math, and then writes back the value again. This is much slower than just directly working on both half values at the same time and I don't think the compiler can fix this.

More generally, I'm not sure adding a half4 struct is a good idea in the first place. I would advise you to just use the builtin half2 struct since it has hardware support and allows you to write code more closely to the metal.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+    half S(0.0);
+    half M(-INFINITY);
+
+    for (int64_t ic = warp_id; ic < ne11; ic += nwraps) {
```
Rewrite to loop with `ic = ic0 + warp_id`.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+
+        __syncthreads();
+
+        if (tiih == 0) {
```
This is bad in terms of performance. Not only is `if {} else {}` very slow on GPUs but also this forces some of the threads in the warp to idle while the instructions for those threads with `tiih == 0` are being resolved.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +5990,364 @@ static  __global__ void im2col_f32_f16(
+            }
+
+            for (int i = 0; i < D4/tph; ++i) {
+                ps4[hiisg*D4 + tph*i + tiih] = __h4fma(ps4[hiisg*D4 + tph*i + tiih], ms0, __h4mul(ps4[sg*(R*D + 32)/4 + hiisg*D4 + tph*i + tiih], ms1));
```
According to NSight Compute this is the slowest part of the kernel. One of the factors is probably the `if (warp_id == 0)` which forces 31/32 warps to wait.

JohannesGaessler reviewed (commented):
The results that I get with this kernel are incorrect. Your top priority should be to fix this since the performance numbers for a kernel that produces incorrect results are not reliable.

Generally speaking, you should only use 16 bit floats in conjunction with half2. The CUDA registers have a size of 32 bits so using 16 bit values/arithmetic is actually *slower* than using 32 bit values/arithmetic.

FSSRepo committed:
use half2 instead half4

FSSRepo committed:
match to metal impl

ggerganov committed:
metal : avoid redundant loads of the attention

ggerganov committed:
metal : scale and mask in matrix form

ggerganov committed:
metal : fix comment

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

ggerganov committed:
llama : avoid ggml_cast, use F32 query

ggerganov committed:
metal : add parallel reduce version (disabled)

FSSRepo committed:
update implementation

FSSRepo committed:
integrate tensor cores

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

FSSRepo committed:
update impl

FSSRepo commented:
@JohannesGaessler Any idea to solve this? I've already included `#include <mma.h>`, but it still doesn't work.

```bash
URCE=600 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml-cuda.obj 
   "C:\proyects\llama.cpp\ggml-cuda.cu"
C:\proyects\llama.cpp\ggml-cuda.cu(6137): error : name followed by "::" must be a class or namespace name [C:\proyects\llama.cpp\build\ggml.vcxproj]
    typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_a;
            ^

C:\proyects\llama.cpp\ggml-cuda.cu(6137): error : expected an identifier [C:\proyects\llama.cpp\build\ggml.vcxproj]
    typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_a;
                                                           ^

C:\proyects\llama.cpp\ggml-cuda.cu(6137): error : expected an identifier [C:\proyects\llama.cpp\build\ggml.vcxproj]
    typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_a;
                                                               ^

C:\proyects\llama.cpp\ggml-cuda.cu(6137): error : expected an identifier [C:\proyects\llama.cpp\build\ggml.vcxproj]
    typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_a;
  ```

JohannesGaessler commented:
The error only seems to occur with cmake so it most likely is due to tensor core code only being defined for compute capability 7.0 or higher. So wrapping the kernel in `#if __CUDA_ARCH__ >= CC_VOLTA` should fix it.

ggerganov committed:
Merge branch 'master' into gg/flash-attn

ggerganov committed:
metal : move output into local memory + optimize

- the result from each simdgroup now stays in the registers
- significantly reduced SRAM usage
- more efficient skipping of -INF blocks
- avoid simdgroup barrier in hot loop
- add comments

ggerganov committed:
metal : add tests, fix scaling, support C > 32

ggerganov committed:
metal : improve precision

ggerganov committed:
ggml : fix f16 mad

ggerganov committed:
Merge branch 'master' into gg/flash-attn

ggerganov committed:
metal : minor

ggerganov committed:
metal : support Q > 8

ggerganov committed:
tests : add ATTN tests

ggerganov committed:
metal : disable buffer allocation logs

ggerganov committed:
tests : more

ggerganov committed:
metal : faster inner loop for C == 32

FSSRepo committed:
fix compiler error

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

ggerganov committed:
metal : fix array initialization

ggerganov committed:
tests : ifdef

ggerganov committed:
Merge branch 'master' into gg/flash-attn

FSSRepo committed:
latest kernel update, wrong values

ggerganov commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+
+                    for (int64_t i = 0; i < D16; ++i) {
+                        half16x16_bT mk; // transposed key
+                        nvcuda::wmma::load_matrix_sync(mk, pk + i*16, nb11/sizeof(half)); // transpose
```
The resulting matrix `mk` should be transposed. Maybe try changing the layout to column/row major?

FSSRepo replied:
See `half16x16_bT` is `typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, half, nvcuda::wmma::row_major> half16x16_bT;`

ggerganov commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                            int msk_row = i / 16;
+                            mqk[j].x[i] = __float2half(scale * __half2float(mqk[j].x[i]) + msk_p[msk_col + msk_row*msk_ne_row]);
+                        }
+                        nvcuda::wmma::store_matrix_sync(ss + 16*j*T + 16*cc, mqk[j], T, nvcuda::wmma::mem_col_major);
```
I could be misunderstanding how these functions work, but I think all `store_matrix_sync` and `load_matrix_sync` should use `mem_row_major` and only the `mk` matrix should load using `mem_col_major`

JohannesGaessler replied:
For unfused general matrix multiplication kernels, to get correct results for `src0` as fragment a and `src1` as fragment b, you have to load `src0`/fragment a as row major, `src1`/fragment b as col major, and you have to store the result as col major. But I did **not** analyze the exact memory layout needed for a fused kernel that does multiple matrix multiplications at once.

ggerganov commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                nvcuda::wmma::load_matrix_sync(mm, ss + 16*j*T + C + 16*j, T);
+
+                for (int64_t i = 0; i < D16; ++i) {
+                    nvcuda::wmma::mma_sync(lo[j][i], mm, zro, lo[j][i]);
```
This looks incorrect: `lo = mm*0 + lo` ?

FSSRepo replied:
It might be wrong, but `lo` it's an accumulator, so I can't use it directly as a `matrix_b`. I created the `zro` matrix filled with zeros since that part of the code `lo` matrices are currently filled with zeros.

JohannesGaessler replied:
I don't understand what you're trying to say but this line is definitely wrong. As Georgi said, this is a no-op that does not change the values of `lo[j][i]` whatsoever.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+            for (int64_t i = 0; i < D16; ++i) {
+                nvcuda::wmma::load_matrix_sync(mq[j][i], sq + 16*j*T + i*16, T);
+            }
+        }
```
Preloading all of these fragments from shared memory to registers will very likely have bad performance because it will cause a lot of register pressure which will then cause the tensor core fragments to spill into VRAM. Also these loops can and should be unrolled. But as I said before: getting a kernel that produces correct results should be your top priority, **only focus on optimizing performance afterwards**.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                half16x16_a ms0;
+                half16x16_a ms1;
+                half16x16_b t;
+                half16x16_acc t2;
```
`t2` is never initialized with `fill_fragment` so this is causing undefined behavior.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                    nvcuda::wmma::mma_sync(t2, ms1, t, t2);
+
+                    // t <- lo
+                    for (uint32_t k = 0; k < t.num_elements; k++) {
+                        t.x[k] = lo[j][i].x[k];
+                    }
```
NVIDIA does **not** give you any guarantees regarding how the data for tensor core fragments is distributed between threads in a warp. So even if the memory layout between the two matrix types happens to be the same on the GPUs that you're testing your code on it may produce incorrect results on other architectures.

For now I would recommend that you replace this with `store_matrix_sync` and `load_matrix_sync` to make sure it's not the reason the results are incorrect. Maybe check the FlashAttention CUDA code to see what they do.

ggerganov commented:
The following patch should fix some of the issues, but the compute is still not correct:

```diff
diff --git a/ggml-cuda.cu b/ggml-cuda.cu
index 5229e15d..dd4d97bc 100644
--- a/ggml-cuda.cu
+++ b/ggml-cuda.cu
@@ -6158,9 +6158,9 @@ static __global__ void flash_attn_f32(
 }
 
 #if __CUDA_ARCH__ >= CC_VOLTA
-typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_a;
-typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_b;
-typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, half, nvcuda::wmma::row_major> half16x16_bT;
+typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a, 16, 16, 16, half, nvcuda::wmma::row_major> half16x16_a;
+typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, half, nvcuda::wmma::row_major> half16x16_b;
+typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_b, 16, 16, 16, half, nvcuda::wmma::col_major> half16x16_bT;
 typedef nvcuda::wmma::fragment<nvcuda::wmma::accumulator, 16, 16, 16, half>                     half16x16_acc;
 
 // based on metal version
@@ -6200,19 +6200,19 @@ static __global__ void flash_attn_ext_f16(
     const int iq2 = blockIdx.y;
     const int iq1 = blockIdx.x * Q;
 
-    const int D2 = D/2;
+    const int D2  = D/2;
     const int D16 = D/16;
     const int Q16 = Q/16;
     const int NW = WARP_SIZE;
-    const int SH = (C + D); // shared memory per simdgroup in (half)
+    const int SH = (C + Q); // shared memory per simdgroup in (half)
 
     const int T  = D + num_warps*SH; // shared memory size per query in (half)
     const int T2 = T/2;              // shared memory size per query in (half2)
 
     extern __shared__  half __flash_attn_f16_shmem[];
     // pq
-    half  * sq  = (half  *) (__flash_attn_f16_shmem +             0*D); // holds the query data
-    half2 * sq2 = (half2 *) (__flash_attn_f16_shmem +             0*D); // same as above but in half2
+    half  * sq  = (half  *) (__flash_attn_f16_shmem +              0*D); // holds the query data
+    half2 * sq2 = (half2 *) (__flash_attn_f16_shmem +              0*D); // same as above but in half2
     half  * ss  = (half  *) (__flash_attn_f16_shmem + warp_id*SH + 1*D); // scratch buffer for attention and diagonal matrix
     half16x16_acc lo[Q16][D16];
 
@@ -6305,7 +6305,7 @@ static __global__ void flash_attn_ext_f16(
 
                     for (int64_t i = 0; i < D16; ++i) {
                         half16x16_bT mk; // transposed key
-                        nvcuda::wmma::load_matrix_sync(mk, pk + i*16, nb11/sizeof(half)); // transpose
+                        nvcuda::wmma::load_matrix_sync(mk, pk + i*16, nb11/sizeof(half));
 
                         for (int64_t j = 0; j < Q16; ++j) {
                             nvcuda::wmma::mma_sync(mqk[j], mq[j][i], mk, mqk[j]);
@@ -6321,7 +6321,7 @@ static __global__ void flash_attn_ext_f16(
                             int msk_row = i / 16;
                             mqk[j].x[i] = __float2half(scale * __half2float(mqk[j].x[i]) + msk_p[msk_col + msk_row*msk_ne_row]);
                         }
-                        nvcuda::wmma::store_matrix_sync(ss + 16*j*T + 16*cc, mqk[j], T, nvcuda::wmma::mem_col_major);
+                        nvcuda::wmma::store_matrix_sync(ss + 16*j*T + 16*cc, mqk[j], T, nvcuda::wmma::mem_row_major);
                     }
                 }
             }
@@ -6393,14 +6393,19 @@ static __global__ void flash_attn_ext_f16(
 
             // O = diag(ms)*O
             for (int64_t j = 0; j < Q16; ++j) {
-                half16x16_a mm;
-                half16x16_b zro;
+                //half16x16_a mm;
+                //half16x16_b mo;
 
-                nvcuda::wmma::fill_fragment(zro, 0.0);
-                nvcuda::wmma::load_matrix_sync(mm, ss + 16*j*T + C + 16*j, T);
+                //nvcuda::wmma::load_matrix_sync(mm, ss + 16*j*T + C + 16*j, T);
 
                 for (int64_t i = 0; i < D16; ++i) {
-                    nvcuda::wmma::mma_sync(lo[j][i], mm, zro, lo[j][i]);
+                    //nvcuda::wmma::load_matrix_sync(mo, lo[j][i]);
+                    //nvcuda::wmma::mma_sync(lo[j][i], mm, mo, lo[j][i]);
+
+                    for (uint32_t k = 0; k < 16*16; k++) {
+                        half tmp = ss[(j + k/16)*T + C + 16*j + k/16];
+                        lo[j][i].x[k] = tmp * lo[j][i].x[k];
+                    }
                 }
             }
 
@@ -6444,7 +6449,7 @@ static __global__ void flash_attn_ext_f16(
         if (warp_id == sg) {
             for (int64_t j = 0; j < Q16; ++j) {
                 for (int64_t i = 0; i < D16; ++i) {
-                    nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_col_major);
+                    nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_row_major);
                 }
             }
         }
@@ -6483,6 +6488,8 @@ static __global__ void flash_attn_ext_f16(
                 half16x16_b t;
                 half16x16_acc t2;
 
+                nvcuda::wmma::fill_fragment(t2, 0.0);
+
                 nvcuda::wmma::load_matrix_sync(ms0, ss + 16*j*T + C + 16*j,         T);
                 nvcuda::wmma::load_matrix_sync(ms1, ss + 16*j*T + C + 16*j + sg*SH, T);
 
@@ -6504,22 +6511,18 @@ static __global__ void flash_attn_ext_f16(
     if (warp_id == 0) {
         for (int64_t j = 0; j < Q16; ++j) {
             for (int64_t i = 0; i < D16; ++i) {
-                nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_col_major);
+                nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_row_major);
             }
         }
     }
 
-    float2 * dst2 = (float2 *) dst;
-
     // final rescale with 1/S and store to global memory
     if (warp_id == 0) {
         for (int64_t j = 0; j < Q && iq1 + j < ne01; ++j) {
             const float S = __half2float(ss[j*T + 0]);
 
-            for (int64_t i = lane_id; i < D2; i += NW) {
-                dst2[(iq3*ne2*ne1 + iq2 + (iq1 + j)*ne1)*D2 + i] =  __half22float2(sq2[j*T2 + i]);
-                dst2[(iq3*ne2*ne1 + iq2 + (iq1 + j)*ne1)*D2 + i].x /= S;
-                dst2[(iq3*ne2*ne1 + iq2 + (iq1 + j)*ne1)*D2 + i].y /= S;
+            for (int64_t i = lane_id; i < D; i += NW) {
+                dst[(iq3*ne2*ne1 + iq2 + (iq1 + j)*ne1)*D + i] = __half2float(sq[j*T + i]) / S;
             }
         }
     }
@@ -10549,13 +10552,14 @@ inline void ggml_cuda_flash_attn_ext(const ggml_tensor * Q, const ggml_tensor *
 
     const int nqpb = 16; // queries per block
     const int ncpw = 32; // cache values per warp (does not work for other values)
-    const int nwarps = Q->ne[1] <= nqpb ? MAX(4, MIN(K->ne[1]/ncpw, 32)) : 4;
+    //const int nwarps = Q->ne[1] <= nqpb ? MAX(4, MIN(K->ne[1]/ncpw, 32)) : 4;
+    const int nwarps = 1;
 
     dim3 blocks_num((Q->ne[1] + nqpb - 1) / nqpb, Q->ne[2], Q->ne[3]);
     dim3 block_dim(32, nwarps, 1);
 
-    int shmem = nqpb*(Q->ne[0] + nwarps*(Q->ne[0] + 1*ncpw))*(sizeof(float)/2);
-    printf("shared memory: %d bytes [%i, %i, %i]\n\n", shmem, Q->ne[0], Q->ne[1], Q->ne[2]);
+    int shmem = nqpb*(Q->ne[0] + nwarps*(ncpw + nqpb))*(sizeof(float)/2);
+    //printf("shared memory: %d bytes [%i, %i, %i]\n\n", shmem, Q->ne[0], Q->ne[1], Q->ne[2]);
     switch (Q->ne[0])
     {
     case 16:
diff --git a/tests/test-backend-ops.cpp b/tests/test-backend-ops.cpp
index f57e8ab1..07182c6d 100644
--- a/tests/test-backend-ops.cpp
+++ b/tests/test-backend-ops.cpp
@@ -1726,6 +1726,7 @@ static bool test_backend(ggml_backend_t backend, test_mode mode, const char * op
     test_cases.emplace_back(new test_pad());
     test_cases.emplace_back(new test_leaky_relu());
 
+#if 0
     for (int hs : { 64,  80,  96, 112, 128, 256, }) {
         for (int nh : { 32, }) {
             for (int kv : { 512, 1024, 2048, 4096, }) {
@@ -1736,6 +1737,18 @@ static bool test_backend(ggml_backend_t backend, test_mode mode, const char * op
             }
         }
     }
+#else
+    for (int hs : { 128, }) {
+        for (int nh : { 32, }) {
+            for (int kv : { 512, 1024, }) {
+                for (int nb : { 1, 2, 4, 8, 512 }) {
+                    test_cases.emplace_back(new test_attn          (hs, nh, kv, nb));
+                    test_cases.emplace_back(new test_flash_attn_ext(hs, nh, kv, nb));
+                }
+            }
+        }
+    }
+#endif
 
 #if !defined(__SANITIZE_THREAD__)
     // FIXME: these tests use too much memory with thread sanitizer
```

Most likely there is still some confusion with the matrix layouts. See @JohannesGaessler comments and try to figure out what are the correct parameters.

Also, for now user just 1 warp in order to avoid debugging the "reduce" part of the kernel at the end. When it works with 1 warp, then you can increase it and debug the rest of the kernel.

ggerganov commented:
I was looking today further into this, and I think we first have to change the mask type from F32 to F16, in order to be able to express `mqk = mqk*scale + mask` in matrix form with `half` fragments. Will try to do this change today

FSSRepo commented:
@ggerganov Apply your changes, and if the results are close enough to the real ones, keep in mind that I conducted the test without considering mask and scale (mask = null, scale = 1.0 (no scale)).

ggerganov commented:
Yes, there is something not right there as it is.

But either way, when we switch to F16 mask, then this code will become very similar to the Metal implementation. And regarding the `lo = diag(ms)*lo` part - we can fix this with a little bit of extra sram to store and load the `lo` accumulator into a `matrix_b` fragment. Then this can also be expressed in matrix form and I suspect the results will then match

FSSRepo commented:
@ggerganov Just out of curiosity, I was wondering why `diag(ms)` is multiplied by `lo`, but `lo` doesn't have previous data when the operation `lo = diag(ms) * lo` is performed.

```c++
// zero out lo
    for (int64_t j = 0; j < Q16; ++j) {
        for (int64_t i = 0; i < D16; ++i) {
            nvcuda::wmma::fill_fragment(lo[j][i], 0.0);
        }
    }
    ...
   // O = diag(ms)*O
   for (int64_t j = 0; j < Q16; ++j) {
                for (int64_t i = 0; i < D16; ++i) {
                    for (uint32_t k = 0; k < 16*16; k++) {
                        half tmp = ss[(16*j + k%16)*T + C + 16*j + k%16];
                        lo[j][i].x[k] = tmp * lo[j][i].x[k];
                    }
                }
            }
```

ggerganov committed:
Merge branch 'master' into gg/flash-attn

ggerganov committed:
ggml : switch to padded F16 mask for ggml_soft_max, ggml_flash_attn_ext

FSSRepo committed:
fix kernel

ggerganov commented:
For the first iteration over the `ic` loop, `lo` is empty, but for all following iterations it is no longer empty.

```c
// this accumulates lo = lo + mv*mk;
nvcuda::wmma::mma_sync(lo[j][i], mv, mk, lo[j][i]);
```

ggerganov commented:
@FSSRepo I've added a commit in the FA PR that changes to F16 mask:

[8ad92dc](https://github.com/ggerganov/llama.cpp/pull/5021/commits/8ad92dc1ec9aa6549c68900daa7ab93b57fa3ae5)

If you pick the commit here, I can try to update the CUDA kernel to do everything in matrix form using F16 fragments.

FSSRepo committed:
fix naive implementation

FSSRepo committed:
Merge branch 'gg/flash-attn' of https://github.com/ggerganov/llama.cpp into flash-attn-cuda

FSSRepo committed:
cuda: mask as fp16

FSSRepo commented:
@ggerganov I already brought in the latest commit from your pull request. I think you can try adding the method for masking here.

ggerganov committed:
cuda : fix flash_attn kernel to produce same results as CPU

ggerganov committed:
cuda : avoid extra QxQ matrix in shared memory

ggerganov committed:
cuda : switch to F16 scalars + tune warps for RTX 2060

ggerganov committed:
cuda : increase C to 128 for better performance

sorasoras commented:
`cmake .. -DLLAMA_CUBLAS=ON -DLLAMA_CUDA_FORCE_DMMV=TRUE -DLLAMA_CUDA_DMMV_X=32 -DLLAMA_CUDA_MMV_Y=2 -DLLAMA_CUDA_F16=TRUE -DBUILD_SHARED_LIBS=TRUE
`
```
.\test-flash-attention.exe
load_model: ggml tensor size    = 368 bytes
load_model: backend buffer size = 0.04 MB
load_model: using CUDA backend
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: Tesla P40, compute capability 6.1, VMM: no
main: compute buffer size: 0.00 MB
computing time: 0.9200 ms

Performing test:
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000 0.0000
```


Could I assumed this is unsupported on p40?

FSSRepo commented:
@sorasoras To use flash attention, the library requires the Tensor Cores, which are present starting from the Volta architecture (Nvidia V100) and newer.

FSSRepo committed:
Merge pull request #3 from ggerganov/flash-attn-cuda

cuda : fix flash_attn kernel to produce same results as CPU

FSSRepo committed:
fix mask nullptr

FSSRepo committed:
don't require LLAMA_CUDA_F16 to compile

ggerganov commented on ggml-cuda.cu:
```diff
@@ -615,10 +657,23 @@ static __device__ __forceinline__ half2 warp_reduce_sum(half2 a) {
 }
 
+static __device__ __forceinline__ half warp_reduce_sum(half x) {
+#ifdef __CUDA_ARCH__ >= CC_VOLTA
```
```suggestion
#if  __CUDA_ARCH__ >= CC_VOLTA
```

FSSRepo committed:
#ifdef -> #if + fix check -inf

FSSRepo committed:
cmake: remove unused changes

FSSRepo commented:
@ggerganov Have you tried conducting perplexity tests with the flash attention kernel? I've been experimenting with phi-2, and it's generating nonsensical text. Unfortunately, I don't have a llama model on my computer for testing, and I understand that the phi-2 model has precision issues.

ggerganov commented:
`phi-2` requires F32 kernel. The `master` implementation also uses F32 precision for the matrix multiplications:

https://github.com/ggerganov/llama.cpp/blob/32de3f57a97268e74a6e4dc1335b28ed7fee08db/llama.cpp#L4578-L4584

For now, test with LLaMA / Mistral models. When and if we get the performance, we will look into `phi-2` support

FSSRepo committed:
refactor flash_attn function + improve tests

ggerganov commented on ggml-cuda.cu:
```diff
@@ -6601,8 +6601,8 @@ static __global__ void flash_attn_ext_f16(
-                    const half ms = __hisinf(m) ? __float2half(0.0f) : hexp(m - M[j]);
-                    const half vs = __hisinf(s) ? __float2half(0.0f) : hexp(s - M[j]);
+                    const half ms = __hisinf(m) == -1 ? __float2half(0.0f) : hexp(m - M[j]);
+                    const half vs = __hisinf(s) == -1 ? __float2half(0.0f) : hexp(s - M[j]);
```
Hm, is this correct? `__hisinf` returns bool:

https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH____HALF__COMPARISON.html#group__CUDA__MATH____HALF__COMPARISON_1gc6698aa5f990114a7bcd129ed9375f17

FSSRepo replied:
![Screenshot_20240202-080354_Chrome.jpg](https://github.com/Pints-App/llama.cpp/assets/57494570/3144ece6-e9d1-424c-988f-2447f21570be)

Returns an int (0 not infinite, 1 positive infinite, -1 negative infinite), here we are detecting a -INFINITE

ggerganov replied:
Oh wow 🤦 - you are correct

JohannesGaessler commented:
On my hardware (1x RTX 3090) the FlashAttention kernel is suffering from tail effects for perplexity calculations with a batch size of 512. You can see this in NSight Compute under "PM Sampling":

![Screenshot_20240203_103132](https://github.com/Pints-App/llama.cpp/assets/18492268/0532c4e3-1911-4378-a567-b9208e4a5e1d)

The number of active warps for the last ~30% of the runtime is very low and this in turn reduces the GPU utilization. For more information you can look at the "Launch Statistics":

![Screenshot_20240203_103248](https://github.com/Pints-App/llama.cpp/assets/18492268/bd7dc4c7-13dc-407c-8ab7-e4a9b85b691a)

The number of waves per SM is 2.08. The actual runtime of a kernel will be proportional to this value rounded up to the next integer. Ideally you want this number to be just below a whole integer so that for all waves the GPU utilization is high. The worse-case scenario is something like 1.01 because then you have to do 2 waves when you could probably do a slightly modified version of the kernel in a single wave which would make it ~2x faster.

However, this is something that you should optimize **at the end**. The number of waves will depend strongly on implementation details, the specific GPU, and the batch size. The number of waves will be proportional to `batch_size/(occupancy*num_streaming_multiprocessors)`. My immediate advice would be to optimize the other aspects of the kernel first but to keep tail effects in mind.

JohannesGaessler committed:
unroll 2 loops, int64_t -> int, 309 µs

JohannesGaessler commented:
I made a PR with some very simple performance optimizations because it took me the same time as it would have been writing Github comments: https://github.com/Pints-App/llama.cpp/pull/4

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5972,19 +6214,593 @@ static  __global__ void im2col_f32_f16(
+        int ne2,
+        int ne3) {
+#if __CUDA_ARCH__ >= CC_VOLTA
+    const int warp_id = threadIdx.y;
+    const int lane_id = threadIdx.x;
+
+    const int num_warps = blockDim.y; // number of warps
+    const int iq3 = blockIdx.z;
+    const int iq2 = blockIdx.y;
+    const int iq1 = blockIdx.x * Q;
```
This code copies the values from `threadIdx` and `blockIdx` to registers. It should be very slightly faster to access them this way. But if register pressure is an issue it would make sense to remove these local variables and to instead use `threadIdx` and `blockIdx` directly.

If possible, make `num_warps` a template parameter so the compiler gets more information and you don't have to use a register for it.

JohannesGaessler reviewed (commented):
This code has a lot of `if(...){...}` statements. Try to avoid these if possible since they are slow on GPUs. If you cannot avoid them, try to provide the compiler with enough information via template parameters to resolve the branching at compile time so the runtime is unaffected.

FSSRepo commented:
@JohannesGaessler 

> PM Sampling

I don't see the PM Sampling section in my Nsight Compute. 

![Screenshot 2024-02-03 101921](https://github.com/Pints-App/llama.cpp/assets/57494570/9b831920-2aad-44b4-b31e-122efb4cb362)
![Screenshot 2024-02-03 101931](https://github.com/Pints-App/llama.cpp/assets/57494570/85b0e20e-6b2d-4152-82b1-79ea76e81b5f)
![Screenshot 2024-02-03 101937](https://github.com/Pints-App/llama.cpp/assets/57494570/035bc9b8-a0a7-41c7-8674-16beb682ed58)

FSSRepo committed:
Merge pull request #4 from Pints-App/jg/flash-attn-cuda

unroll 2 loops, int64_t -> int, 309 µs

JohannesGaessler commented:
Under Connect -> Metrics I just ticked "full" and then I got it.

FSSRepo committed:
bring optimizations from gg/flash-attn

FSSRepo committed:
Merge branch 'flash-attn-cuda' of https://github.com/Pints-App/llama.cpp into flash-attn-cuda

FSSRepo committed:
fix merge conflicts

FSSRepo commented:
> Under Connect -> Metrics I just ticked "full" and then I got it.

I don't see that option; it must be because my toolkit is outdated.

JohannesGaessler commented:
At a batch size of 1 the size of the CUDA grid for the `flash_attn_ext_f16` kernel is extremely small:

![Screenshot_20240211_094822](https://github.com/Pints-App/llama.cpp/assets/18492268/1b6b8e90-c1f9-462f-bf69-41d0ffa40750)

As a consequence both the memory and compute throughput are low. At a batch size of 512 the grid is larger but it is still very small:

![Screenshot_20240211_095221](https://github.com/Pints-App/llama.cpp/assets/18492268/a3e0f865-5a08-46bb-acc2-54036481f6db)

There are only 1.25 waves per SM and as a consequence ~40% of the runtime is lost to tail effects:

![Screenshot_20240211_095449](https://github.com/Pints-App/llama.cpp/assets/18492268/1477cb77-8dbc-4c9e-9533-a0170209fea3)

If at all possible, the kernel should be rewritten in such a way that the work is distributed among more CUDA blocks or with more warps per CUDA block.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -5972,19 +6214,599 @@ static  __global__ void im2col_f32_f16(
+                nvcuda::wmma::load_matrix_sync(ms1, ss + 16*j*T + C + 16*j + sg*SH, T);
+
+                for (int i = 0; i < D16; ++i) {
+                    nvcuda::wmma::load_matrix_sync(t, sq + 16*j*T + i*16, T);
+                    nvcuda::wmma::mma_sync(t2, ms1, t, zr);
+
+                    // convert accumulator to matrix_b
+                    nvcuda::wmma::store_matrix_sync(   sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_row_major);
+                    nvcuda::wmma::load_matrix_sync (t, sq + 16*j*T + i*16, T);
+
+                    nvcuda::wmma::mma_sync(lo[j][i], ms0, t, t2);
+                }
+            }
+        }
     }
+
+    // store result to shared memory (reuse sq)
+    if (warp_id == 0) {
+        for (int j = 0; j < Q16; ++j) {
+            for (int i = 0; i < D16; ++i) {
+                nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_row_major);
+            }
+        }
+    }
+
+    // final rescale with 1/S and store to global memory
+    if (warp_id == 0) {
+        for (int j = 0; j < Q && iq1 + j < ne01; ++j) {
+            const half S = ss[j*T + 0];
+
+            for (int i0 = 0; i0 < D; i0 += NW) {
+                const int i = i0 + lane_id;
+                if (i >= D) {
+                    break;
+                }
+
+                dst[(iq3*ne2*ne1 + iq2 + (iq1 + j)*ne1)*D + i] = __half2float(sq[j*T + i] / S);
+            }
+        }
+    }
```
My guess based on just static code analysis is that this is the part of the code that is causing issues when you increase the number of warps. But since I am not able to increase the number of warps (and still get correct results) this could be very wrong.

ggerganov replied:
I think it works correctly for up to 8 warps - after that it runs out of shared memory

JohannesGaessler replied:
Okay, but how do I set the number of warps? If I just set e.g. `nwarps = 4` on line 10940 I get NaN with 

```
export model_name=llama_2-7b && export quantization=q4_0
./perplexity --n-gpu-layers 99 --model models/nvme/${model_name}-${quantization}.gguf --file wikitext-2-raw/wiki.test.raw --mlock --chunks 1
```

>after that it runs out of shared memory

It's implemented in a convoluted and tedious way but it is possible to raise the shared memory limit for CUDA kernels so that a single block can use 100% of the shared memory of an SM. For example:

https://github.com/JohannesGaessler/llama.cpp/blob/979a9bf1bef96fbeafd3a0a605c642ab098021ee/ggml-cuda.cu#L6723-L6735

ggerganov replied:
I'm not sure about the state of this branch. It works on my branch:

https://github.com/ggerganov/llama.cpp/tree/gg/flash-attn

```diff
diff --git a/ggml-cuda.cu b/ggml-cuda.cu
index 713a6a89..99925259 100644
--- a/ggml-cuda.cu
+++ b/ggml-cuda.cu
@@ -10932,7 +10932,7 @@ inline void ggml_cuda_flash_attn_ext(const ggml_tensor * Q, const ggml_tensor *
 
     const int nwarps_max = 8; // TODO: we don't want to launch too much warps. how much is too much?
                               // TODO: produces wrong results for nwarps > 8 (RTX 2060) - not sure why
-    const int nwarps = Q->ne[1] <= nqpb ? std::max(2, std::min((int) K->ne[1]/ncpw, nwarps_max)) : 1;
+    const int nwarps = Q->ne[1] <= nqpb ? std::max(2, std::min((int) K->ne[1]/ncpw, nwarps_max)) : 4;
 
     dim3 blocks_num((Q->ne[1] + nqpb - 1) / nqpb, Q->ne[2], Q->ne[3]);
     dim3 block_dim(32, nwarps, 1);
 ```

```bash
LLAMA_CUBLAS=1 make -j && ./perplexity -m ./models/llama-7b-v2/ggml-model-q4_0.gguf -f ./build-cublas/wikitext-2-raw/wiki.test.raw -ngl 99
```

```
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 0.36 seconds per pass - ETA 3.90 minutes
[1]4.3401,[2]4.8393,[3]5.4619,[4]6.0629,[5]6.1965,[6]6.0945,[7]6.2706,^C
```
